### PR TITLE
feat: Support for experiment names

### PR DESCRIPTION
**Pull request by Raalsky**

### CODE REVIEW

File: src/neptune_fetcher/read_only_run.py [hunk 1]
```diff
@@ -54,22 +54,34 @@ def get_attribute_value_from_entry(entry: TableEntry, name: str) -> Optional[str
...
                 container_id=QualifiedName(f"CUSTOM/{read_only_project.project_identifier}/{custom_id}"),
                 expected_container_type=None,
             )
             self.with_id = experiment.sys_id
+        elif experiment_name is not None:
+            experiment = read_only_project.fetch_experiments_df(
+                query=f"`sys/name`:string = '{experiment_name}'", limit=1, columns=["sys/id"]
+            )
+            if len(experiment) == 0:
+                raise ValueError(f"No experiment found with name '{experiment_name}'")
```
Comment (AleksanderWWW) [inline src/neptune_fetcher/read_only_run.py:83]: Why do we raise an exception if no experiment is found, but don't do it if custom run id doesn't match any run/experiment?

File: src/neptune_fetcher/read_only_run.py [hunk 2]
```diff
@@ -54,22 +54,34 @@ def get_attribute_value_from_entry(entry: TableEntry, name: str) -> Optional[str
...
+        if with_id is None and custom_id is None and experiment_name is None:
+            raise ValueError("Either `with_id`, `custom_id` or `experiment_name` must be provided.")
 
         if custom_id is not None:
             experiment = read_only_project._backend.get_metadata_container(
                 container_id=QualifiedName(f"CUSTOM/{read_only_project.project_identifier}/{custom_id}"),
                 expected_container_type=None,
             )
             self.with_id = experiment.sys_id
+        elif experiment_name is not None:
```
Comment (AleksanderWWW) [inline src/neptune_fetcher/read_only_run.py:78]: What will happen if someone provides both a custom id and an experiment name, but they don't match with each other? Like custom id from experiment1 and name from experiment2?

File: src/neptune_fetcher/read_only_run.py [hunk 3]
```diff
@@ -54,22 +54,34 @@ def get_attribute_value_from_entry(entry: TableEntry, name: str) -> Optional[str
...
         self.project = read_only_project
 
         verify_type("with_id", with_id, (str, type(None)))
         verify_type("custom_id", custom_id, (str, type(None)))
+        verify_type("experiment_name", experiment_name, (str, type(None)))
 
-        if with_id is None and custom_id is None:
-            raise ValueError("Either `with_id` or `custom_id` must be provided.")
+        if with_id is None and custom_id is None and experiment_name is None:
+            raise ValueError("Either `with_id`, `custom_id` or `experiment_name` must be provided.")
```
Comment (kgodlewski) [inline src/neptune_fetcher/read_only_run.py:70]:
```suggestion
            raise ValueError("Either `with_id`, `custom_id`, or `experiment_name` must be provided.")
```

File: src/neptune_fetcher/read_only_run.py (hunk 2 above)
Comment (kgodlewski) [reply src/neptune_fetcher/read_only_run.py:78]: I think those arguments should be made mutually exclusive to avoid confusion

**Review by kgodlewski — DISMISSED** [review]
I would address Aleksander's comments, looks good otherwise

File: src/neptune_fetcher/read_only_run.py (hunk 2 above)
Comment (Raalsky) [reply src/neptune_fetcher/read_only_run.py:85]: You're right. Done

File: src/neptune_fetcher/read_only_run.py (hunk 1 above)
Comment (Raalsky) [reply src/neptune_fetcher/read_only_run.py:90]: Sure, done

File: src/neptune_fetcher/read_only_run.py [hunk 4]
```diff
@@ -54,24 +55,50 @@ def get_attribute_value_from_entry(entry: TableEntry, name: str) -> Optional[str
...
         verify_type("custom_id", custom_id, (str, type(None)))
+        verify_type("experiment_name", experiment_name, (str, type(None)))
+
+        if with_id is None and custom_id is None and experiment_name is None:
+            raise ValueError("Either `with_id`, `custom_id`, or `experiment_name` must be provided.")
 
-        if with_id is None and custom_id is None:
-            raise ValueError("Either `with_id` or `custom_id` must be provided.")
+        if sum([with_id is not None, custom_id is not None, experiment_name is not None]) > 1:
+            raise ValueError("Only one of `with_id`, `custom_id` or `experiment_name` can be provided.")
```
Comment (AleksanderWWW) [inline src/neptune_fetcher/read_only_run.py:74]: I think that those two conditions can be merged into `if sum... != 1` raise with a message like
'exactly one of ... must be specified`

What do you think?

Comment (AleksanderWWW) [reply src/neptune_fetcher/read_only_run.py:74]: Also we can substitute `param is not None` with `bool(param)` to also catch a situation when an empty string is passed (which should also be forbidden as far as I can tell)

Comment (Raalsky) [reply src/neptune_fetcher/read_only_run.py]: Let's keep `None`, I'll apply the `!= 1`. Done

**Review by AleksanderWWW — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -379,11 +402,12 @@ Can be created
...
-|---------------------|-------------------|---------|------------------------------------------------------------------|
-| `read_only_project` | `ReadOnlyProject` | -       | Source project from which run will be fetched.                   |
-| `with_id`           | `Optional[str]`   | `None`  | ID of the Neptune run to fetch. Example: `RUN-1`. Exclusive with the `custom_id` parameter. |
-| `custom_id`         | `Optional[str]`   | `None`  | Custom ID of the Neptune run to fetch. Example: `high_albattani`. Exclusive with the `with_id` parameter. |
+| Name                | Type              | Default | Description                                                                                                                      |
+|---------------------|-------------------|---------|----------------------------------------------------------------------------------------------------------------------------------|
+| `read_only_project` | `ReadOnlyProject` | -       | Source project from which run will be fetched.                                                                                   |
```
Comment (normandy7) [inline README.md:407]:
```suggestion
| `read_only_project` | `ReadOnlyProject` | -       | Project from which the run is fetched.                                                                                   |
```

File: src/neptune_fetcher/read_only_project.py [hunk 6]
```diff
@@ -161,6 +161,20 @@ def fetch_read_only_runs(
...
+        self,
+        names: Optional[List[str]] = None,
+    ) -> Iterator[ReadOnlyRun]:
+        """Lists experiments of the project in the form of read-only runs.
+
+        Returns a generator of `ReadOnlyRun` instances.
+
+        Args:
+            names: List of experiment names to fetch.
+        """
```
Comment (normandy7) [inline src/neptune_fetcher/read_only_project.py:174]:
```suggestion
        Example:
            ```
            project = ReadOnlyProject()
            for run in project.fetch_read_only_experiments(names=["yolo-v2", "yolo-v3"]):
                ...
            ```
        """
```

File: src/neptune_fetcher/read_only_run.py [hunk 7]
```diff
@@ -54,24 +55,50 @@ def get_attribute_value_from_entry(entry: TableEntry, name: str) -> Optional[str
...
+        experiment_name: Optional[str] = None,
     ) -> None:
         self.project = read_only_project
 
         verify_type("with_id", with_id, (str, type(None)))
         verify_type("custom_id", custom_id, (str, type(None)))
+        verify_type("experiment_name", experiment_name, (str, type(None)))
+
+        if with_id is None and custom_id is None and experiment_name is None:
+            raise ValueError("Either `with_id`, `custom_id`, or `experiment_name` must be provided.")
```
Comment (normandy7) [inline src/neptune_fetcher/read_only_run.py:71]:
```suggestion
            raise ValueError("You must provide one of: `with_id`, `custom_id`, and `experiment_name`.")
```

File: src/neptune_fetcher/read_only_run.py [hunk 8]
```diff
@@ -54,24 +55,50 @@ def get_attribute_value_from_entry(entry: TableEntry, name: str) -> Optional[str
...
 
-        if with_id is None and custom_id is None:
-            raise ValueError("Either `with_id` or `custom_id` must be provided.")
+        if sum([with_id is not None, custom_id is not None, experiment_name is not None]) != 1:
+            raise ValueError("Exactly one of `with_id`, `custom_id` or `experiment_name` can be provided.")
```
Comment (normandy7) [inline src/neptune_fetcher/read_only_run.py:74]:
```suggestion
            raise ValueError("You must provide exactly one of: `with_id`, `custom_id`, and `experiment_name`.")
```

File: src/neptune_fetcher/read_only_project.py [hunk 9]
```diff
@@ -161,6 +161,26 @@ def fetch_read_only_runs(
...
+    def fetch_read_only_experiments(
+        self,
+        names: Optional[List[str]] = None,
+    ) -> Iterator[ReadOnlyRun]:
+        """Lists experiments of the project in the form of read-only runs.
+
+        Returns a generator of `ReadOnlyRun` instances.
+
+        Args:
+            names: List of experiment names to fetch.
```
Comment (normandy7) [inline src/neptune_fetcher/read_only_project.py:173]:
```suggestion
            names: List of experiment names to fetch.

```

**Review by normandy7 — APPROVED** [review]

**Pull request merged**